### PR TITLE
SB-21731 feat: Certificate pre processor

### PR DESCRIPTION
**Pull request by PradyumnaNagendra**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-21731" title="SB-21731" target="_blank">StorySB-21731</a>  Migration to Flink Job & Refactoring - Certificate Generation job 
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/collectioncert/functions/IssueCertificateHelper.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        val enrolledUser: EnrolledUser = validateEnrolmentCriteria(event, criteria.getOrElse(config.enrollment, Map[String, AnyRef]()).asInstanceOf[Map[String, AnyRef]], certName)(metrics, cassandraUtil, config)
+        //validateAssessmentCriteria
+        val assessedUser = validateAssessmentCriteria(event, criteria.getOrElse(config.assessment, Map[String, AnyRef]()).asInstanceOf[Map[String, AnyRef]], enrolledUser.userId)(metrics, cassandraUtil, config)
+        //validateUserCriteria
+        val userDetails = validateUser(assessedUser, criteria.getOrElse(config.user, Map[String, AnyRef]()).asInstanceOf[Map[String, AnyRef]])(metrics, config, httpUtil)
+        
+        //generateCertificateEvent
+        if(!userDetails.isEmpty) {
+            generateCertificateEvent(event, template, userDetails, enrolledUser, certName)(metrics, config, cache, httpUtil)
+        } else throw new Exception(s"""User :: ${event.userId} did not match the criteria for batch :: ${event.batchId} and course :: ${event.courseId}""")
```
Comment (maheshkumargangula) [inline credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/collectioncert/functions/IssueCertificateHelper.scala:167]: Throwing exception is not correct.

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+        
+        
+    }
+
+    def isValidEvent(event: Event): Boolean = {
+        config.issueCertificate.equalsIgnoreCase(event.action) && !event.batchId.isEmpty && !event.courseId.isEmpty && 
+          !event.userId.isEmpty
+    }
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala]: Done

File: credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                                context: ProcessFunction[Event, String]#Context,
+                                metrics: Metrics): Unit = {
+        try {
+            metrics.incCounter(config.totalEventsCount)
+            if(isValidEvent(event)) {
+                val certTemplates = fetchTemplates(event)(metrics)
+                if(!certTemplates.isEmpty) {
+                    certTemplates.map(template => {
+                        val certEvent = issueCertificate(event, template._2)(cassandraUtil, cache, metrics, config, httpUtil)
+                        println(certEvent)
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala]: removed

File: credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            } else {
+                logger.info(s"Invalid request : ${event}")
+                metrics.incCounter(config.skippedEventCount)
+            }
+        } catch {
+            case ex: Exception => {
+                logger.error(s"Certificate generate event failed sent to next topic for event : ${event}", ex)
+                metrics.incCounter(config.failedEventCount)
+                context.output(config.failedEventOutputTag, ScalaJsonUtil.serialize(event.eData))
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/functions/CollectionCertPreProcessorFn.scala]: Done

File: credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/collectioncert/functions/IssueCertificateHelper.scala (hunk 1 above)
Comment (PradyumnaNagendra) [reply credential-generator/collection-cert-pre-processor/src/main/scala/org/sunbird/job/collectioncert/functions/IssueCertificateHelper.scala]: replaced with logger

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**